pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -208,10 +211,109 @@ func (n *nodeHealthMap) set(name string, data *nodeHealthData) {
 	n.nodeHealths[name] = data
 }
 
+type podUpdateItem struct {
+	namespace string
+	name      string
+}
+
+type evictionStatus int
+
+const (
+	unmarked = iota
+	toBeEvicted
+	evicted
+)
+
+type nodeEviction struct {
+	status        evictionStatus
+	pods          map[types.UID]*v1.Pod
+	deletionStamp time.Time
+}
+
+// nodeEvictionMap is NOT thread-safe.
+type nodeEvictionMap struct {
+	lock          sync.Mutex
+	nodeEvictions map[string]*nodeEviction
+}
+
+func newNodeEvictionMap() *nodeEvictionMap {
+	return &nodeEvictionMap{
+		nodeEvictions: make(map[string]*nodeEviction),
+	}
+}
+
+func (n *nodeEvictionMap) setStatus(nodeName string, status evictionStatus) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.addNodeEvictionIfNeeded(nodeName)
+	n.nodeEvictions[nodeName].deletionStamp = time.Time{}
+	n.nodeEvictions[nodeName].status = status
+}
+
+func (n *nodeEvictionMap) getStatus(nodeName string) evictionStatus {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.addNodeEvictionIfNeeded(nodeName)
+	return n.nodeEvictions[nodeName].status
+}
+
+func (n *nodeEvictionMap) addPods(nodeName string, pods []*v1.Pod) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.addNodeEvictionIfNeeded(nodeName)
+	for i := range pods {
+		n.nodeEvictions[nodeName].pods[pods[i].UID] = pods[i]
+	}
+}
+
+func (n *nodeEvictionMap) deletePods(nodeName string) []*v1.Pod {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.addNodeEvictionIfNeeded(nodeName)
+	pods := n.nodeEvictions[nodeName].pods
+	n.nodeEvictions[nodeName].pods = make(map[types.UID]*v1.Pod)
+	podList := make([]*v1.Pod, 0, len(pods))
+	for _, v := range pods {
+		podList = append(podList, v)
+	}
+	return podList
+}
+
+func (n *nodeEvictionMap) markForDeletion(nodeName string) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	if _, exists := n.nodeEvictions[nodeName]; exists {
+		n.nodeEvictions[nodeName].deletionStamp = time.Now()
+	}
+}
+
+func (n *nodeEvictionMap) trimDeletedNodeEvictions() {
```

Comment:
Do we really need this complicated logic?

Once the node gets removed, all the pods from that node will be handled by podgc controller (gc-orphaned).
So why can't we simply remove the entry for our node immediately after the node is being deleted?